File: WdtcCore/src/main/java/org/wdt/wdtc/core/auth/yggdrasil/YggdrasilTextures.java
package org.wdt.wdtc.core.auth.yggdrasil;

import com.google.gson.annotations.SerializedName;
import lombok.Getter;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.SkinUtils;
import org.wdt.wdtc.core.utils.URLUtils;

import java.io.File;
import java.io.IOException;
import java.net.URL;

public class YggdrasilTextures {
  private final String username;
  private String url;

  public YggdrasilTextures(YggdrasilAccounts yggdrasilAccounts) {
    username = yggdrasilAccounts.getUserName();
    url = yggdrasilAccounts.getUrl();
  }

  public void setUrl(String url) {
    this.url = url;
  }

  public String getUserJsonUrl() {
    return url + "/csl/" + username + ".json";
  }

  public String getUserSkinHash() throws IOException {
    Csl csl = getCsl();
    Skins skins = csl.getSkins();
    return skins.getSkinKind();
  }

  public File getUserSkinFile() {
    return new SkinUtils(username).getSkinFile();
  }

  public void DownloadUserSkin() throws IOException {
    String UserSkinHash = getUserSkinHash();
    File SkinPath = new File(FileManger.getMinecraftComSkin(), UserSkinHash.substring(0, 2) + "/" + UserSkinHash);
    DownloadUtils.StartDownloadTask(getSkinUrl(), SkinPath);
  }

  public URL getSkinUrl() throws IOException {
    return new URL(url + "/textures/" + getUserSkinHash());
  }

  public Csl getCsl() throws IOException {
    return JsonObjectUtils.parseObject(URLUtils.getURLToString(getUserJsonUrl()), Csl.class);
  }

  public SkinUtils getUtils() throws IOException {
    SkinUtils utils = new SkinUtils(getUserSkinFile());
    utils.setUserSkinInput(getSkinUrl().openStream());
    return utils;
  }


  public static class Csl {
    @SerializedName("username")
    private String UserName;
    @Getter
    @SerializedName("skins")
    private Skins Skins;
    @SerializedName("cape")
    private String Cape;

  }

  @Getter
  public static class Skins {
    @SerializedName(value = "default", alternate = {"slim"})
    private String SkinKind;
  }

}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/auth/yggdrasil/YggdrasilTextures.kt
package org.wdt.wdtc.core.auth.yggdrasil

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.parseObject
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.manger.isWindows
import org.wdt.wdtc.core.manger.minecraftComSkin
import org.wdt.wdtc.core.utils.*
import java.io.File
import java.net.URL

class YggdrasilTextures(
	private val userName: String,
	private val url: String,
) {
	
	private val userSkinHash: String
		get() = "$url/csl/$userName.json".toURL().toStrings().parseObject<Csl>().skins?.skinKind.noNull()
	
	
	suspend fun startDownloadUserSkin() {
		if (!isWindows) return
		val skinPath = userSkinHash.let {
			File(minecraftComSkin, "${it.substring(0, 2)}/$it")
		}
		startDownloadTask(skinUrl to skinPath)
	}
	
	private val skinUrl: URL
		get() = URL("$url/textures/${userSkinHash}")
	
	val utils: SkinUtils
		get() = SkinUtils(SkinUtils.getUserSkinFile(userName)).apply {
			userSkinInput = skinUrl.newInputStream()
		}
	
	class Csl {
		@SerializedName("username")
		private val userName: String? = null
		
		@SerializedName("skins")
		val skins: Skins? = null
		
		@SerializedName("cape")
		val cape: String? = null
	}
	
	class Skins {
		@SerializedName(value = "default", alternate = ["slim"])
		val skinKind: String? = null
	}
}
